tests/expecter.go

```diff
@@ -0,0 +1,351 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package tests
+
+import (
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+
+	expect "github.com/google/goexpect"
+	"google.golang.org/grpc/codes"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/util/net/dns"
+)
+
+const (
+	PromptExpression = `(\$ |\# )`
+	CRLF             = "\r\n"
+)
+
+var (
+	shellSuccess = regexp.MustCompile(RetValue("0"))
+	shellFail    = regexp.MustCompile(RetValue("[1-9].*"))
+)
+
+func VmiConsoleExpectBatch(vmi *v1.VirtualMachineInstance, expected []expect.Batcher, timeout time.Duration) error {
+	virtClient, err := kubecli.GetKubevirtClient()
+	if err != nil {
+		return err
+	}
+
+	expecter, _, err := NewConsoleExpecter(virtClient, vmi, 30*time.Second)
+	if err != nil {
+		return err
+	}
+	defer expecter.Close()
+
+	resp, err := expecter.ExpectBatch(expected, timeout)
+	if err != nil {
+		log.DefaultLogger().Object(vmi).Infof("%v", resp)
+	}
+	return err
+}
+
+func VmiConsoleRunCommand(vmi *v1.VirtualMachineInstance, command string, timeout time.Duration) error {
+	err := VmiConsoleExpectBatch(vmi, []expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: command + "\n"},
+		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: "echo $?\n"},
+		&expect.BCas{C: []expect.Caser{
+			&expect.Case{
+				R: shellSuccess,
+				T: expect.OK(),
+			},
+			&expect.Case{
+				R: shellFail,
+				T: expect.Fail(expect.NewStatus(codes.Unavailable, command+" failed")),
+			},
+		}},
+	}, timeout)
+	if err != nil {
+		return fmt.Errorf("Failed to ping VMI %s, error: %v", vmi.Name, err)
+	}
+	return nil
+}
+
+func LoggedInCirrosExpecter(vmi *v1.VirtualMachineInstance) (expect.Expecter, error) {
```

Comment:
I think these Log-In helpers are different from the others here. These are equivalent to the ping or other specific operations that use expecters, but are not generic.

Reply:
maybe we can move them to `login.go` ? do something like `console/login.go`, `console/expecter.go` ?

Reply:
I would move these to `login.go` for now and not try to aggregate them into a "console" package at the moment.

I agree that the expecter is an implementation detail, but I am not sure if it is limited to a console (or what a console is exactly). Lets start with a login.go and then think about organizing them in a package as a later step.

Reply:
Done.